sdk/storage/src/blobs/block_blob_client.cpp

```diff
@@ -98,23 +98,18 @@ namespace Azure { namespace Storage { namespace Blobs {
     BlobRestClient::BlockBlob::StageBlockFromUriOptions protocolLayerOptions;
     protocolLayerOptions.BlockId = blockId;
     protocolLayerOptions.SourceUri = sourceUri;
-    if (options.SourceOffset != std::numeric_limits<decltype(options.SourceOffset)>::max())
+    if (options.SourceOffset.HasValue() && options.SourceLength.HasValue())
     {
-      if (options.SourceLength == 0)
-      {
-        protocolLayerOptions.SourceRange = std::make_pair(
-            options.SourceOffset, std::numeric_limits<decltype(options.SourceOffset)>::max());
-      }
-      else
-      {
-        protocolLayerOptions.SourceRange
-            = std::make_pair(options.SourceOffset, options.SourceOffset + options.SourceLength - 1);
-      }
+      protocolLayerOptions.SourceRange = std::make_pair(
+          options.SourceOffset.GetValue(),
+          options.SourceOffset.GetValue() + options.SourceLength.GetValue() - 1);
     }
-    else
+    else if (options.SourceOffset.HasValue())
     {
-      protocolLayerOptions.SourceRange
-          = std::make_pair(std::numeric_limits<uint64_t>::max(), uint64_t(0));
+      protocolLayerOptions.SourceRange = std::make_pair(
+          options.SourceOffset.GetValue(),
+          std::numeric_limits<
```

Comment:
A range of "0-" is valid right? Why we need to add this here?

Reply:
https://docs.microsoft.com/en-us/rest/api/storageservices/specifying-the-range-header-for-blob-service-operations#format-1-bytesstartbyte-

Reply:
Ditto for other range end checks.

Reply:
What do you mean?

Reply:
If `SourceLength` is null, we can send `x-ms-range: SourceOffset-` leaving the end to be empty. Why are we enforcing max value here?

Reply:
Because in protocol layer it's a `std::pair`

Reply:
Then the `std::pair<int64_t, int64_t>` should be changed to `std::pair<int64_t, Azure::Core::Nullable<int64_t>>` right?

Reply:
maybe it should be changed to `std::pair<Azure::Core::Nullable<int64_t>, Azure::Core::Nullable<int64_t>>`

Reply:
I got the feeling that Nullable is abused here.

Reply:
https://github.com/Azure/azure-sdk-for-cpp/issues/232